ports/zephyr/machine_timer.c

```diff
@@ -0,0 +1,200 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2025 Daniel Campora on behalf of REMOTE TECH LTD
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdint.h>
+#include <stdio.h>
+
+#include "py/mperrno.h"
+#include "py/obj.h"
+#include "py/runtime.h"
+#include "modmachine.h"
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/sys/printk.h>
+
+#define TIMER_MS_PER_TICK (1000)
+
+typedef enum _machine_timer_mode_t {
+    TIMER_MODE_ONE_SHOT = 0,
+    TIMER_MODE_PERIODIC
+} machine_timer_mode_t;
+
+typedef struct _machine_timer_obj_t {
+    mp_obj_base_t base;
+
+    struct k_timer my_timer;
+
+    machine_timer_mode_t mode;
+    uint32_t period_ms;
+
+    mp_obj_t callback;
+
+    struct _machine_timer_obj_t *next;
+} machine_timer_obj_t;
+
+const mp_obj_type_t machine_timer_type;
+
+static mp_obj_t machine_timer_init_helper(machine_timer_obj_t *self, mp_uint_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args);
+
+static void machine_timer_callback(struct k_timer *timer) {
+    machine_timer_obj_t *self = (machine_timer_obj_t *)k_timer_user_data_get(timer);
+    if (self->mode == TIMER_MODE_ONE_SHOT) {
+        k_timer_stop(&self->my_timer);
```

Comment:
I think call `machine_timer_deinit(MP_OBJ_FROM_PTR(self))` here, so that the one-shot timer is removed from the linked list.

That way you should be able to create one-shot timers indefinitely without deinit'ing them explicitly.

Reply:
Yeah that's a good point. Done now. Thanks.

Reply:
Great thanks.  I tested on `frdm_k64f` creating many one-shot timers in a loop, and it works.